test/chainrules.jl

```diff
@@ -1,15 +1,35 @@
 @testset "test rrule" begin
-    @testset "mul" begin
+    @testset "matrix multiplication" begin
         # sparse * vector
-        A = sprand(5, 3, 0.5)
-        B = rand(3)
-        test_rrule(*, A ⊢ sprand_tangent(A), B; check_thunked_output_tangent=false)
-        # test_rrule(*, A ⊢ sprand_tangent(A), B) # FIXME: Thunk doesn't yet supported.
+        #for T in (Float64)
+            A = sprand(5, 3, 0.5)
+            B = rand(3)
+            test_rrule(*, A ⊢ sprand_tangent(A), B; check_thunked_output_tangent=false)
+            # test_rrule(*, A ⊢ sprand_tangent(A), B) # FIXME: Thunk doesn't yet supported.
 
-        # sparse * dense matrix
-        A = sprand(5, 3, 0.5)
-        B = rand(3, 2)
-        test_rrule(*, A ⊢ sprand_tangent(A), B; check_thunked_output_tangent=false)
-        # test_rrule(*, A ⊢ sprand_tangent(A), B) # FIXME: Thunk doesn't yet supported.
+            # sparse * dense matrix
+            A = sprand(5, 3, 0.5)
+            B = rand(3, 2)
+            test_rrule(*, A ⊢ sprand_tangent(A), B; check_thunked_output_tangent=false)
+            # test_rrule(*, A ⊢ sprand_tangent(A), B) # FIXME: Thunk doesn't yet supported.
+        #end 
+    end
+
+    @testset "dense matrix-sparse matrix multiplication" begin
+        # dense matrix * sparse matrix
+        #for T in (Float64)
+            B = rand(10, 10)
+            A = sprand(10, 5, 0.5)
+            test_rrule(*, B, A ⊢ sprand_tangent(A); check_thunked_output_tangent=false)
+        #end 
+    end
```

Comment:
I think maybe complex case should be added into rrule. But it really confused me, when I tried to write as follows:
``` julia
for T in (Float64)
            B = rand(T, 10, 10)
            A = sprand(T, 10, 5, 0.5)
            test_rrule(*, B, A ⊢ sprand_tangent(A); check_thunked_output_tangent=false)
end 
```
It went wrong and told me
``` julia
MethodError: no method matching iterate(::Nothing)
```
Why? Is there any differences?

Reply:
In julia 1.7, the error message is more intuitive to understand:
```julia
ERROR: MethodError: no method matching iterate(::Type{Float64})
Closest candidates are:
  iterate(::Union{LinRange, StepRangeLen}) at range.jl:737
  iterate(::Union{LinRange, StepRangeLen}, ::Int64) at range.jl:737
  iterate(::T) where T<:Union{Base.KeySet{<:Any, <:Dict}, Base.ValueIterator{<:Dict}} at dict.jl:695
  ...
Stacktrace:
 [1] top-level scope
   @ ./REPL[6]:1
```

`(Float64)` This is a type, `(Float64,)` is a tuple.

Reply:
Thx!!!

Reply:
After I corrected the syntax error,  I guess may be I need to add to_vec for complex case when I see the error message:

``` julia
Got exception outside of a @test
  MethodError: no method matching _j′vp(::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, ::ComposedFunction{ComposedFunction{ComposedFunction{typeof(first), typeof(FiniteDifferences.to_vec)}, FiniteDifferences.var"#79#80"{ChainRulesTestUtils.var"#fnew#42"{ChainRulesTestUtils.var"#call#50"{NamedTuple{(), Tuple{}}}, Tuple{typeof(*), SparseMatrixCSC{ComplexF64, Int64}, Matrix{ComplexF64}}, Tuple{Bool, Bool, Bool}}}}, FiniteDifferences.var"#Tuple_from_vec#48"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{var"#SparseMatrixCSC_from_vec#1"{SparseMatrixCSC{ComplexF64, Int64}, SparseMatrixCSC{ComplexF64, Int64}}, FiniteDifferences.var"#Array_from_vec#34"{Matrix{ComplexF64}, FiniteDifferences.var"#Vector_from_vec#32"{Vector{ComplexF64}, Vector{FiniteDifferences.var"#Complex_from_vec#21"}, Vector{Vector{Float64}}}}}}}, ::Vector{Float64}, ::Vector{ComplexF64})
  Closest candidates are:
    _j′vp(::Any, ::Any, ::Vector{var"#s16"} where var"#s16"<:Real, ::Vector{var"#s15"} where var"#s15"<:Real) at /Users/lijie/.julia/packages/FiniteDifferences/W3rQO/src/grad.jl:78
  Stacktrace:
```